internal/db/p2p_replicator.go

```diff
@@ -335,3 +344,377 @@ func (db *db) loadAndPublishReplicators(ctx context.Context) error {
+			db.retryReplicators(ctx)
+		}
+	}
+}
+
+func (db *db) handleReplicatorFailure(ctx context.Context, r event.ReplicatorFailure) error {
+	err := db.updateReplicatorStatus(ctx, r.PeerID.String(), false)
+	if err != nil {
+		return err
+	}
+	err = db.createIfNotExistsReplicatorRetry(ctx, r.PeerID.String())
+	if err != nil {
+		return err
+	}
+	docIDKey := core.NewReplicatorRetryDocIDKey(r.PeerID.String(), r.DocID)
+	err = db.Peerstore().Put(ctx, docIDKey.ToDS(), []byte{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (db *db) handleCompletedReplicatorRetry(ctx context.Context, r retryStatus) error {
+	if r.Success {
+		done, err := db.deleteReplicatorRetryIfNoMoreDocs(ctx, r.PeerID)
+		if err != nil {
+			return err
+		}
+		if done {
+			err := db.updateReplicatorStatus(ctx, r.PeerID, true)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		err := db.setReplicatorNextRetry(ctx, r.PeerID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// updateReplicatorStatus updates the status of a replicator in the peerstore.
+func (db *db) updateReplicatorStatus(ctx context.Context, peerID string, active bool) error {
+	key := core.NewReplicatorKey(peerID)
+	repBytes, err := db.Peerstore().Get(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+	rep := client.Replicator{}
+	err = json.Unmarshal(repBytes, &rep)
+	if err != nil {
+		return err
+	}
+	switch active {
+	case true:
+		rep.Status = client.ReplicatorStatusActive
+		if rep.Status == client.ReplicatorStatusInactive {
+			rep.LastStatusChange = time.Time{}
+		}
+	case false:
+		rep.Status = client.ReplicatorStatusInactive
+		if rep.Status == client.ReplicatorStatusActive {
+			rep.LastStatusChange = time.Now()
+		}
+	}
+	b, err := json.Marshal(rep)
+	if err != nil {
+		return err
+	}
+	return db.Peerstore().Put(ctx, key.ToDS(), b)
+}
+
+type retryInfo struct {
+	NextRetry  time.Time
+	NumRetries int
+	Retrying   bool
+}
+
+func (db *db) createIfNotExistsReplicatorRetry(ctx context.Context, peerID string) error {
+	key := core.NewReplicatorRetryIDKey(peerID)
+	exists, err := db.Peerstore().Has(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+	if !exists {
+		r := retryInfo{
+			NextRetry:  time.Now().Add(db.retryIntervals[0]),
+			NumRetries: 0,
+		}
+		b, err := cbor.Marshal(r)
+		if err != nil {
+			return err
+		}
+		err = db.Peerstore().Put(ctx, key.ToDS(), b)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	return nil
+}
+
+func (db *db) retryReplicators(ctx context.Context) {
+	q := query.Query{
+		Prefix: core.REPLICATOR_RETRY_ID,
+	}
+	results, err := db.Peerstore().Query(ctx, q)
+	if err != nil {
+		log.ErrorContextE(ctx, "Failed to query replicator retries", err)
+		return
+	}
+	defer closeQueryResults(results)
+	now := time.Now()
+	for result := range results.Next() {
+		rInfo := retryInfo{}
+		err = cbor.Unmarshal(result.Value, &rInfo)
+		if err != nil {
+			log.ErrorContextE(ctx, "Failed to unmarshal replicator retry info", err)
```

Comment:
thought: If this error is ever hit, it seems likely that it will not be the only record (programming error).  I worry that if we do not delete the record in this block we will have a very rapidly growing database and log file.

Reply:
✅